### PR TITLE
[WIP] Introduction of linalg, orthonormalization, and statistic utilities for symmetric random variables.

### PR DESCRIPTION
**Pull request by Danfoa**

### CODE REVIEW

**Comment by alekfrohlich** [comment]
What about nn/symmetric/(...) then stuff? We could have the exact same dir structure there (linalg, modules, etc.) and than lazy import that inside nn via __init__ logic.

So far we don't have a statistics.py (perhaps we should? Whitening, for instance, is in linalg.py).

**Comment by Danfoa** [comment]
I feel strongly we should separate. In my case I have a ton of symmetry aware statistic functions that are not lin algebra:

Linalg:
- Orthogonal projections to subspaces.
- Disentangled representation

Statistics:
- mean, var, cov, cross_cov of symmetric random variables

File: linear_operator_learning/nn/utils/statistics.py
Comment (Danfoa) [inline linear_operator_learning/nn/utils/statistics.py:1]: Also where would you put these Unbiased estimations of ||Cxy||_F^2. 

Seems rather specific for linalgebra things

**Comment by alekfrohlich** [comment]
> I feel strongly we should separate. In my case I have a ton of symmetry aware statistic functions that are not lin algebra:
> 
> Linalg:
> 
> * Orthogonal projections to subspaces.
> * Disentangled representation
> 
> Statistics:
> 
> * mean, var, cov, cross_cov of symmetric random variables

For me it is fine to separate.

File: linear_operator_learning/nn/utils/statistics.py
Comment (alekfrohlich) [reply linear_operator_learning/nn/utils/statistics.py:1]: > Also where would you put these Unbiased estimations of ||Cxy||_F^2.
> 
> Seems rather specific for linalgebra things

what about symm_orthn_fro_reg? We have orthn_fro_reg and orthn_logfro_reg inside functional.py. These are regularizer functions that we make available also via the module interface in modules/loss.py (_RegularizedLoss base class).

**Comment by Danfoa** [comment]
Done @alekfrohlich @pietronvll have a look at the functions introduced. 

We need to solve the optinal dependency [`symm_torch`](https://github.com/Danfoa/symm_torch) which is the lightweight package where I put all the symmetric stuff, that is too specific to put here. 

Check the `nn/stats.py` and the `nn/symmetric/stats.py` and `nn/symmetric/linalg.py` files. 

The rest of the changes listed in this pr come from the previour pr #7 (because I branch from Alek's branch), once that PR is merged, this one should follow.

**Comment by alekfrohlich** [comment]
Sounds good, should we added it to which uv group? A new one called symm? To examples?

Regarding the files, why break the conventions we had for orthonormality regularization? Also why the general cross_cov? Please read my older comment. If you'd like to propose changes to the loss/reg interface, then perhaps it would be better to first finish this PR and then change all the losses and regularizers to match accordingly in a new PR (otherwise we will go 3 PRs deep!).

File: linear_operator_learning/nn/symmetric/stats.py [hunk 1]
```diff
@@ -0,0 +1,386 @@
+"""Statistics utilities for symmetric random variables with known group representations."""
+
+import numpy as np
+import torch
+from escnn.group import Representation
+from symm_torch.utils.rep_theory import isotypic_decomp_rep
+
+
+def symmetric_moments(x: torch.Tensor, rep_X: Representation) -> [torch.Tensor, torch.Tensor]:
```
Comment (pietronvll) [inline linear_operator_learning/nn/symmetric/stats.py:9]: I propose to match — where possilbe — the signature and API of [the following torch function](https://pytorch.org/docs/stable/generated/torch.var_mean.html), and rename it simply `symmetric_var_mean`.

**Review by pietronvll — COMMENTED** [review]
@Danfoa please change the typing suggestion for torch tensors from `torch.Tensor` to simply `Tensor`, to match the convention we used throughout

**Comment by pietronvll** [comment]
> Done @alekfrohlich @pietronvll have a look at the functions introduced.
> 
> We need to solve the optinal dependency [`symm_torch`](https://github.com/Danfoa/symm_torch) which is the lightweight package where I put all the symmetric stuff, that is too specific to put here. I would list it as an optional dependency, and ask the user to install it whenever anything from `nn/symmetric` is imported. Does this sound ok?
> 
> Check the `nn/stats.py` and the `nn/symmetric/stats.py` and `nn/symmetric/linalg.py` files.
> 
> The rest of the changes listed in this pr come from the previour pr #7 (because I branch from Alek's branch), once that PR is merged, this one should follow.

I'm slightly skeptical of this solution:

1. By not having tests, any changes in `symm_torch` can break our `symmetric` submodule without us noticing. I noticed @Danfoa already added some tests. To be a little more on the safe side, we might add tests to our version-bumping workflow. 

2. As we agreed, we would like the code to be as portable and hackable as possible. How much do we depend upon `symm_torch` and `escnn`? Are they stable dependencies? Can they introduce breaking changes to our code? I totally understand we don't want to re-invent the wheel to do symmetric stuff. I'm just pushing to have as few failure points as possible.

**Comment by pietronvll** [comment]
> > Done @alekfrohlich @pietronvll have a look at the functions introduced.
> > We need to solve the optinal dependency [`symm_torch`](https://github.com/Danfoa/symm_torch) which is the lightweight package where I put all the symmetric stuff, that is too specific to put here. I would list it as an optional dependency, and ask the user to install it whenever anything from `nn/symmetric` is imported. Does this sound ok?
> > Check the `nn/stats.py` and the `nn/symmetric/stats.py` and `nn/symmetric/linalg.py` files.
> > The rest of the changes listed in this pr come from the previour pr #7 (because I branch from Alek's branch), once that PR is merged, this one should follow.
> 
> I'm slightly skeptical of this solution:
> 
> 1. By not having tests, any changes in `symm_torch` can break our `symmetric` submodule without us noticing. I noticed @Danfoa already added some tests. To be a little more on the safe side, we might add tests to our version-bumping workflow.
> 2. As we agreed, we would like the code to be as portable and hackable as possible. How much do we depend upon `symm_torch` and `escnn`? Are they stable dependencies? Can they introduce breaking changes to our code? I totally understand we don't want to re-invent the wheel to do symmetric stuff. I'm just pushing to have as few failure points as possible.

As a follow-up, I was reading about [`uv` workspaces](https://docs.astral.sh/uv/concepts/projects/workspaces), which are designed to maintain separate projects sharing a common enforced set of dependencies. The `symm_torch` from @Danfoa, as well as `kooplearn` itself can be members of a workspace whose root is linear operator learning. 

This has the possible disadvantage that we have to maintain a monorepo with all of these dependencies. Any thoughts @Danfoa, @alekfrohlich?

**Comment by pietronvll** [comment]
@Danfoa if you plan to upload symm_torch on pypi we can just add some integration tests to run before merging into main

File: linear_operator_learning/nn/symmetric/linalg.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        if irrep_id == rep_X.group.trivial_representation.id:
+            # this dimension is associated with a trivial irrep
+            S[irrep_dims, irrep_dims] = 1
+        cum_dim += irrep.size
+
+    inv_projector = Qx @ S @ Qx_T
+    return inv_projector
+
+
+def isotypic_signal2irreducible_subspaces(X: torch.Tensor, repX: Representation):
```
Comment (Danfoa) [inline linear_operator_learning/nn/symmetric/linalg.py:49]: import as Tensor

File: linear_operator_learning/nn/symmetric/stats.py [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+import numpy as np
+import torch
+from escnn.group import Representation
+from symm_torch.utils.rep_theory import isotypic_decomp_rep
+from torch import Tensor
+
+from linear_operator_learning.nn.symmetric.linalg import invariant_orthogonal_projector
+
+
+def symmetric_moments(x: Tensor, rep_X: Representation) -> [Tensor, Tensor]:
```
Comment (Danfoa) [inline linear_operator_learning/nn/symmetric/stats.py:12]: torch name for same fn

File: linear_operator_learning/nn/symmetric/stats.py [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+    """Compute the mean and variance of a symmetric random variable.
+
+    Args:
+        x: (Tensor) of shape (N, dx) containing the observations of the symmetric random variable
```
Comment (Danfoa) [inline linear_operator_learning/nn/symmetric/stats.py:16]: Shape section

File: linear_operator_learning/nn/symmetric/stats.py [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+    mean = torch.einsum("ij,...j->...i", P_inv, mean_empirical)
+
+    # Symmetry constrained variance computation.
+    Cx = cross_cov(x, x, rep_X, rep_X)
+    var = torch.diag(Cx)
+    return mean, var
+
+
+def isotypic_cross_cov(
+    X: Tensor,
```
Comment (Danfoa) [inline linear_operator_learning/nn/symmetric/stats.py:40]: lowecaps

File: linear_operator_learning/nn/symmetric/stats.py [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+    where :math:`d = \text{dim(irrep)}` and :math:`\mathbf{D}_{xy} \in \mathbb{R}^{m_x \times m_y}` and :math:`\mathbf{C}_{yx} \in \mathbb{R}^{(m_x \cdot d) \times (m_y \cdot d)}`.
+
+    Being :math:`m_x` and :math:`m_y` the multiplicities of the irrep in X and Y respectively. This implies that the matrix :math:`\mathbf{D}_{xy}`
+    represents the free parameters of the cross-covariance we are required to estimate. To do so we reshape
+    the signals :math:`X \in \mathbb{R}^{n \times (m_x \cdot d)}` and :math:`Y \in \mathbb{R}^{n \times (m_y \cdot d)}` to :math:`X_{\text{sing}} \in \mathbb{R}^{(d \cdot n) \times m_x}` and :math:`Y_{\text{sing}} \in \mathbb{R}^{(d \cdot n) \times m_y}`
+    respectively. Ensuring all dimensions of the irreducible subspaces associated to each multiplicity of the irrep are
+    considered as a single dimension for estimating :math:`\mathbf{D}_{xy} = \frac{1}{n \cdot d} X_{\text{sing}}^T Y_{\text{sing}}`.
+
+    Args:
+        X (Tensor): shape (..., n, m_x \cdot d) where n is the number of samples and m_x the multiplicity of the irrep in X.
```
Comment (Danfoa) [inline linear_operator_learning/nn/symmetric/stats.py:63]: Shape sec

File: linear_operator_learning/nn/symmetric/stats.py [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+    else:
+        Cxy = Cxy_iso
+
+    if not y_in_iso_basis:
+        Cxy = Cxy @ Qx_T
+
+    return Cxy, Dxy
+
+
+def cross_cov(X: Tensor, Y: Tensor, rep_X: Representation, rep_Y: Representation):
```
Comment (Danfoa) [inline linear_operator_learning/nn/symmetric/stats.py:150]: -> covariance

**Review by Danfoa — COMMENTED** [review]
This also lowers the lowest python version to 3.10
